Add ModuleIndexMerger

sgallagher opened the pull request:
This is the final major piece for libmodulemd 2.0. It adds an object and API for combining the ModuleIndexes retrieved from separate repositories and merging them together into a single view of the system, throwing errors if they encounter unresolvable conflicts. See the modulemd-module-index-merger.h header (or the generated documentation for it) for details.

This series of patches builds upon https://github.com/fedora-modularity/libmodulemd/pull/153 (so it includes the patches from that PR as well).

There are several fixes for issues discovered while writing the Merger. They have all been split into their own patches so they can be reviewed more easily.

These patches also include some basic tests for the merge functionality. Many more will be forthcoming, but those can be written in parallel with the review, so I don't want to hold up what will likely be a very involved code-review for that.

puiterwijk commented on modulemd/v2/meson.build:
```diff
@@ -227,6 +227,25 @@ test('module_index_v2_debug', test_v2_module_index,
      env : test_release_env)
 
+
+# -- Test Modulemd.ModuleIndex (C) -- #
```
`Modulemd.IndexMerger`

puiterwijk commented on modulemd/v2/tests/ModulemdTests/merger.py:
```diff
@@ -0,0 +1,177 @@
+        # Use randomly-selected high and low values to make sure we don't have
+        # sorting issues.
+        merger = Modulemd.ModuleIndexMerger()
+        merger.associate_index(base_index, random.randint(1, 100))
```
Perhaps it's an idea to print these indexes, so that reproducing the test is simple?

sgallagher replied:
I can do that, sure. When I get around to writing the matching test in C, the GLib test library provides a mechanism for doing this as well (see the example in the 1.x code).

puiterwijk replied:
In Python unittests, you can just `print()` them, and they'll only show if the test failed.

sgallagher commented on modulemd/v2/modulemd-module-index.c:
```diff
@@ -750,6 +752,150 @@ modulemd_module_index_add_translation (ModulemdModuleIndex *self,
+
+          if (!current_translation ||
+              modulemd_translation_get_modified (translation) >
+                modulemd_translation_get_modified (current_translation))
```
There's no merge logic for translations. Unlike Defaults, their `modified` field is authoritative. It would be a change in expected behavior from 1.x to support overrides allowing them to go "backwards". If you think that should be reconsidered, please open an issue on it.

sgallagher commented:
Rebased this branch atop the approved PR #153 and added fixup commits for review comments. If approved, I'll squash them into the appropriate places and then merge both PRs together.

puiterwijk commented:
LGTM
